### PR TITLE
feat(gres)!: add mappings for managing `GRESName` and `GRESNode` objects

### LINKED ISSUE #39: Support for multiple lines with same NodeName in gres.conf
### Bug Description

`gresconfig` seems to assume one `NodeName` line per node, limiting configuration options.

### To Reproduce

Attempting to use two `NodeName` entries in `gres.conf` to configure a node `juju-abc654-1` with two distinct types of GPU. Desired state of `gres.conf`:

```
NodeName=juju-abc654-1 Name=gpu Type=tesla_t4 File=/dev/nvidia[0-1]
NodeName=juju-abc654-1 Name=gpu Type=l40s File=/dev/nvidia[2-3]
```

with `slurmutils`:

```python3
from slurmutils.editors import gresconfig
from slurmutils.models import GRESNode

with gresconfig.edit("gres.conf") as config:
    node = GRESNode(
        NodeName="juju-abc654-1",
        Name="gpu",
        Type="tesla_t4",
        File="/dev/nvidia[0-1]",
    )
    node2 = GRESNode(
        NodeName="juju-abc654-1",
        Name="gpu",
        Type="l40s",
        File="/dev/nvidia[2-3]",
    )
    
    config.nodes.update(node.dict())
    # How is node2 added?
```

Edit: further reproducer, using existing `gres.conf` in desired state:

```python3
from slurmutils.editors import gresconfig
from slurmutils.models import GRESNode

with gresconfig.edit("gres.conf") as config:
    print(config.nodes)
```

returns `{'juju-abc654-1': {'Name': 'gpu', 'Type': 'l40s', 'File': '/dev/nvidia[2-3]'}}`, without the `tesla_t4` entry. 

### Environment

Running locally with the latest main commit: https://github.com/charmed-hpc/slurmutils/commit/60bc8e617d4745773995ea55288851bdee28ad6e.

### Relevant log output

```shell
N/A
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by NucciTheBoss**
This PR adds the `GRESNameMapping` and `GRESNodeMapping` objects to make it easier to manage _gres.conf_ data model objects. Rather than throw lists and dictionaries out there all loosey goosey for `GRESConfig`, instead these new maps help facilitate interacting with the configured generic resources and easily share them between the slurmctld and slurmd operators. Couple key changes:

* There's now stricter validation for the _gres.conf_ related models by using `jsonschema`. Helps make sure the input we're putting into the maps are valid.
* Methods for encoding and decoding dictionaries with nested data models. Makes it easier to validate with `jsonschema` and send "across the wire," but still enable us to take advantage of the data models and the descriptors + methods that they provide.
  * The `expand` function is an anamorphism that recursively expands a dictionary with nested data model objects. This makes it easier to punch objects into a JSON schema validator. Custom object hooks are used with the JSON decoder to parse the data models back into Python objects.
* I introduced a `BaseMapping` object to provide methods necessary for `GRESNameMapping` and `GRESNodeMapping`, but in the future I could see `BaseMapping` becoming the new `BaseModel`. The classes do similar things, but `BaseMapping` is more flexible and uses JSON schemas to validate the data models.  

### Breaking changes 

The `names` and `nodes` properties from `GRESConfig` now return `GRESNameMapping` and `GRESNodeMapping` objects respectively rather than `list` and `dict` types.

### Related issues

* Fixes #39 

### Misc.

* I updated the annotations for some functions to be compliant with recommendations set forth by Python 3.10. Big thing is that the many of the types provided by the `typing` module are now deprecated, so their imports are now updated. Another thing is the inclusion of the `typing-extensions` module. This is an official module maintained by the Python team. I bring it in as dependency since it provides the `Self` type which wasn't introduced into Python until 3.11. Since `py310` is still a supported Python version, figured it doesn't hurt to bring `typing-extensions` in. 

  `MutableMapping` and `Sequence` are the recommend annotations for arguments that are either type `dict` or `list` respectively.

### CODE REVIEW

**Review by dsloanm — APPROVED** [review]
Looks good and solves the issue I was hitting!

I've not spotted the stricter validation, however. I accidentally generated a few invalid `gres.conf` files when testing since I can get away with things like: 

```python
$ python
Python 3.12.3 (main, Nov  6 2024, 18:32:19) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from slurmutils.editors import gresconfig
>>> with gresconfig.edit("gres-testing.conf") as config:
...   config.nodes["123"] = [456]
... 
>>> 
$ cat gres-testing.conf
456
$
```

File: README.md [hunk 1]
```diff
@@ -94,22 +94,22 @@ from slurmutils.editors import gresconfig
...
         Name="gpu",
         Type="epyc",
         File="/dev/amd[0-3]",
         Count="12G",
     )
     config.auto_detect = "rsmi"
-    config.names.append(name.dict())
-    config.nodes.updaten(node.dict())
+    config.names[new_gres.name] = [new_gres]
+    config.nodes[new_node.node_name] = [new_node]
```
[the remark this answers is not in the document]
Comment (NucciTheBoss) [reply README.md:112]: Nah, `new_node.node_name` descriptor maps to the value of the `NodeName` field  in _gres.conf_. If you have a line like `NodeName=juju-abc654-1 Name=gpu Type=tesla_t4 File=/dev/nvidia[0-1]`, `new_node.node_name` will return `"juju-abc654-1"`. That value can then be used to get the list of `GRESNode` objects that share the same `NodeName` value.

**Comment by NucciTheBoss** [comment]
> I've not spotted the stricter validation, however. I accidentally generated a few invalid `gres.conf` files when testing since I can get away with things like:

Ah, that's because the validation is only done when the mapping object is first created. If you tried to initialize a new `GRESNodeMapping` with that dictionary, it would "explode in your face":

```pycon
>>> from slurmutils.models import GRESNodeMapping
>>> GRESNodeMapping({"123": [456]})
Traceback (most recent call last):
  File "/home/nucci/git/slurmutils/slurmutils/models/model.py", line 252, in __init__
    validate(d, schema=self._schema)
  File "/home/nucci/.cache/pypoetry/virtualenvs/slurmutils-d46I0P7a-py3.12/lib/python3.12/site-packages/jsonschema/validators.py", line 1332, in validate
    raise error
jsonschema.exceptions.ValidationError: 456 is not of type 'object'
Failed validating 'type' in schema['patternProperties']['^.+$']['items']:
    {'$schema': 'https://json-schema.org/draft/2020-12/schema',
     'type': 'object',
     'properties': {'NodeName': {'type': 'string'},
                    'AutoDetect': {'type': 'string'},
                    'Count': {'type': 'string'},
                    'Cores': {'type': 'array',
                              'items': {'type': 'string'},
                              'uniqueItems': True},
                    'File': {'type': 'string'},
                    'Flags': {'type': 'array',
                              'items': {'type': 'string'},
                              'uniqueItems': True},
                    'Links': {'type': 'array', 'items': {'type': 'string'}},
                    'MultipleFiles': {'type': 'string'},
                    'Name': {'type': 'string'},
                    'Type': {'type': 'string'}},
     'additionalProperties': False}
On instance['123'][0]:
    456
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/home/nucci/git/slurmutils/slurmutils/models/model.py", line 255, in __init__
    raise ModelError(e.message)
slurmutils.exceptions.ModelError: 456 is not of type 'object'
```

Validation is performed on when the mappings are created &mdash; I still need to extend it to the other data models &mdash; but I'd like to eventually get it where validation is also performed before the object is serialized such as when we're writing back out to a configuration file or converting to a JSON string for "sending across the wire" . That's for a future PR however, otherwise the diff for this PR would become massive :sweat_smile:

**Comment by NucciTheBoss** [comment]
v0.11.0 is now published to PyPI: https://pypi.org/project/slurmutils/

**Pull request merged**